livekit/src/room/participant/local_participant.rs

```diff
@@ -15,27 +15,43 @@
 use std::{
     collections::HashMap,
     fmt::Debug,
-    sync::{self, Arc},
+    panic::{self, AssertUnwindSafe},
+    pin::Pin,
+    sync::Arc,
     time::Duration,
 };
 
-use chrono::{TimeZone, Utc};
-
-use libwebrtc::{native::create_random_uuid, rtp_parameters::RtpEncodingParameters};
-use livekit_api::signal_client::SignalError;
-use livekit_protocol as proto;
-use livekit_runtime::timeout;
-use parking_lot::Mutex;
-use proto::request_response::Reason;
-
 use super::{ConnectionQuality, ParticipantInner, ParticipantKind};
+use crate::room::proto::RpcError as RpcError_Proto;
 use crate::{
     e2ee::EncryptionType,
     options::{self, compute_video_encodings, video_layers_from_encodings, TrackPublishOptions},
     prelude::*,
+    room::participant::rpc::{RpcError, RpcErrorCode, MAX_PAYLOAD_BYTES},
     rtc_engine::{EngineError, RtcEngine},
     ChatMessage, DataPacket, SipDTMF, Transcription,
 };
+use chrono::{TimeZone, Utc};
+use futures_util::Future;
+
+use libwebrtc::{native::create_random_uuid, rtp_parameters::RtpEncodingParameters};
+use livekit_api::signal_client::SignalError;
+use livekit_protocol as proto;
+use livekit_runtime::timeout;
+use parking_lot::Mutex;
+use proto::request_response::Reason;
+use tokio::sync::oneshot;
+
+type RpcHandler = Arc<
+    dyn Fn(
+            String,              // request_id
+            ParticipantIdentity, // caller_identity
+            String,              // payload
+            Duration,            // response_timeout_ms
+        ) -> Pin<Box<dyn Future<Output = Result<String, RpcError>> + Send>>
+        + Send
+        + Sync,
+>;
```

Comment:
I see that we expect it to returns a Future but where are we awaiting it?

Also Send + Sync seems like a hard requirement

Reply:
here https://github.com/livekit/rust-sdks/pull/461/files#diff-0e573daa2222454b9b896dc8d914c5a6bb4e1bf442f2200407531a1cb4ff7b7bR813

but i went back over the implementation and found that the catch_unwind isn't actually doing anything anyways, and the example works correctly without it. removing it makes this part clearer, the handler is now awaited directly `handler(...).await`

https://github.com/livekit/rust-sdks/pull/461/files#diff-0e573daa2222454b9b896dc8d914c5a6bb4e1bf442f2200407531a1cb4ff7b7bR804

Reply:
(send + sync discussed offline and we decided to leave it as-is)